src/haptic_feedback.c

```diff
@@ -0,0 +1,183 @@
+        } else {
+            haptic_feedback->haptic_feedback_state = haptic_feedback->error;
+            switch (state->sat) {
+            case SAT_PB_TEMPERATURE:
+                // Careful
+                if (haptic_feedback->haptic_feedback_state.peroid > 0) {
+                    haptic_feedback->haptic_feedback_state.peroid /= 2;
+                }
+                __attribute__((fallthrough));
+            case SAT_PB_LOW_VOLTAGE:
+                __attribute__((fallthrough));
+            case SAT_PB_HIGH_VOLTAGE:
+                if (haptic_feedback->counter == 3) {
+                    haptic_feedback->haptic_feedback_state.peroid *= 2;
+                } else if (haptic_feedback->counter == 4) {
+                    haptic_feedback->counter = 0;
+                }
+                break;
+            default:
+                break;
+            }
+        }
+        // Reset state if a different type of event is coming in
+        if (haptic_feedback->haptic_feedback_state.type != prev) {
+            haptic_feedback_reset(haptic_feedback, current_time);
+        }
+
+        if (!haptic_feedback->tone_in_progress) {
+            haptic_feedback_play_two_tones(haptic_feedback);
+            haptic_feedback->alt_timer = current_time;
+            haptic_feedback->tone_in_progress = true;
+            haptic_feedback->note_duration = 0.3;
+        }
+        haptic_feedback->timer = current_time;
+    }
+}
+
+void haptic_feedback_update(HapticFeedback *haptic_feedback, float current_time) {
+    if (!(VESC_IF->foc_play_tone && VESC_IF->foc_stop_audio)) {
+        return;
+    }
+
+    if (haptic_feedback->tone_in_progress) {
+        if ((fabsf(haptic_feedback->timer - current_time) > haptic_feedback->note_duration) ||
+            (haptic_feedback->haptic_feedback_state.type == HAPTIC_FEEDBACK_BEEP &&
+             haptic_feedback->beep_num == 0)) {
+            haptic_feedback_reset(haptic_feedback, current_time);
+        } else if ((haptic_feedback->haptic_feedback_state.peroid > 0) &&
+                   (fabsf(haptic_feedback->alt_timer - current_time) >
+                    haptic_feedback->haptic_feedback_state.peroid / 1000.0)) {
+            haptic_feedback->alt_timer = current_time;
+            // update for special modes
+            haptic_feedback->counter++;
+            if (haptic_feedback->counter % 2 == 0) {
+                if (haptic_feedback->haptic_feedback_state.type == HAPTIC_FEEDBACK_BEEP) {
+                    haptic_feedback_play_single_tone(
+                        haptic_feedback->haptic_feedback_state.freq,
+                        haptic_feedback->haptic_feedback_state.volt,
+                        0
+                    );
+                    haptic_feedback->beep_num--;
+                } else {
+                    haptic_feedback_play_two_tones(haptic_feedback);
+                }
+            } else {
+                VESC_IF->foc_stop_audio(false);
+            }
+        }
+    }
```

Comment:
This logic is very complex and opaque. You also repeatedly call the play functions all the time, which is not very clean. What I'd do instead is take a similar approach to when you implement an animation. I would store a `bool` that says whether you are currently playing a tone. Then, depending on the time, determine whether you're supposed to be playing a tone, and if that differs from the current state, start or stop playing the tone.

For a duty tone pattern, tone-gap-tone-gap-..., you can do e.g.:
```C
const float tone_length = 0.2f;
const float period = tone_length  * 2; // note: according to convention I call period the whole segment that repeats
time = fmodf(time, period); // now time is your time within a single period
bool should_be_playing = time < tone_length;
```

For lets say three beeps and a pause:
```C
const float tone_length = 0.2f;
const float period = tone_length  * 8; // tone-gap-tone-gap-tone-gap-gap-gap = 8 beats
time = fmodf(time, period);
int beat = floorf(time / tone_length);
bool should_be_playing = (beat % 2) == 0 && beat != 6;
```

I hope this makes sense to you, if not let me know and I'll try to come up with a PoC to demonstrate better.

Reply:
So I don't call any of the functions redundantly, only every period when alt_timer expires. Regarding the example for duty tone pattern, I'm not sure that this logic would blend smoothly with how I'm implementing creating a solid tone for when we've reached the duty cycle threshold (by continue to update alt_timer so it never reaches its next period). If you can provide an alternative that's fine, but I'm not sure I want to spend more cycles here rewriting this.